# Coursera capstone week 3- Segmenting and Clustering Neighborhoods in Toronto


## 3.Explore Neighborhoods in Toronto

First, we need to import all the dependencies and our data set that we prepared in previous part

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
# pd.set_option('display.max_columns', None)
# pd.set_option('display.max_rows', None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


#### Import our data

In [2]:
neighborhoods = pd.read_csv('final_df.csv', index_col = 0)
neighborhoods.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

#### Use geopy library to get the latitude and longitude values of Toronto.

In [5]:
# Create a recursive function to avoid timeout error -> request location until get it
def do_geocode(address):
    try:
        return Nominatim().geocode(address)
    except :
        return do_geocode(address)

In [6]:
address = 'Toronto, Ontario'

location = do_geocode(address)
latitude = location.latitude
longitude = location.longitude
print ('The geographical coordinates of Toronto are {}, {}'.format(latitude, longitude))

C:\ProgramData\Anaconda3\lib\site-packages\geopy\geocoders\osm.py:143: UserWarning: Using Nominatim with the default "geopy/1.17.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  UserWarning


The geographical coordinates of Toronto are 43.653963, -79.387207


#### Create a map of Toronto with neighborhoods superimposed on top.

In [7]:
#create map of Toronto using latitude and longitude
map_toronto = folium.Map(location = [latitude, longitude], zoom_start = 10)

#add markers to the map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'],neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{} - {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
        [lat,lng],
        radius = 5,
        popup = label,
        color = 'blue',
        fill = True,
        fill_color = '#3186cc',
        fill_opacity = 0.7
    ).add_to(map_toronto)

In [8]:
map_toronto

Now, we only want to work with boroughs that contain the word Toronto. Let's create a new dataframe

In [9]:
toronto_data = neighborhoods[neighborhoods['Borough'].apply(lambda x : 'Toronto' in x)]
toronto_data = toronto_data.reset_index(drop = True )
toronto_data.head()

PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

Create a visualization of all the boroughs that contain the word Toronto

In [10]:
map_toronto1 = folium.Map(location = [latitude,longitude], zoom_start = 10)
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{} - {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
        [lat,lng],
        popup = label,
        radius = 5,
        color = 'blue',
        fill = True,
        fill_color = '#3186cc',
        fill_opacity = 0.7,
    ).add_to(map_toronto1)

In [11]:
map_toronto1

Use foursquare to explore the area

In [12]:
CLIENT_ID = '4BJ2KXFXKWWLQED01IF0EHQ5X1QWM04L2MZXVFIYCM1DJ0TD' # your Foursquare ID
CLIENT_SECRET = 'HFBJI3NKE4VI5LCQWIIWHO1AT2BN3BVD1510JKQH5VISP2MW' # your Foursquare Secret
VERSION = '20181025' # Foursquare API version
LIMIT = 100
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 4BJ2KXFXKWWLQED01IF0EHQ5X1QWM04L2MZXVFIYCM1DJ0TD
CLIENT_SECRET:HFBJI3NKE4VI5LCQWIIWHO1AT2BN3BVD1510JKQH5VISP2MW


In [13]:
toronto_data.loc[0, 'Neighborhood']

'The Beaches'

In [14]:
neighborhood_lat = toronto_data.loc[0, 'Latitude']
neighborhood_lon = toronto_data.loc[0, 'Longitude']
neighborhood_name = toronto_data.loc[0, 'Neighborhood']
print ('Latitude and longitude of neighborhood {} are {}, {}'.format(neighborhood_name, 
                                                                     neighborhood_lat, 
                                                                     neighborhood_lon))

Latitude and longitude of neighborhood The Beaches are 43.67635739999999, -79.2930312


### Let's get the top 100 venues in The Beaches within a radius of 500 meters

In [15]:
radius = 500

In [16]:
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID,
    CLIENT_SECRET,
    VERSION,
    neighborhood_lat,
    neighborhood_lon,
    radius,
    LIMIT
)

In [17]:
response = requests.get(url).json()
response

{'meta': {'code': 200, 'requestId': '5bd3d440351e3d1692f063e2'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'The Beaches',
  'headerFullLocation': 'The Beaches, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 3,
  'suggestedBounds': {'ne': {'lat': 43.680857404499996,
    'lng': -79.28682091449052},
   'sw': {'lat': 43.67185739549999, 'lng': -79.29924148550948}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e77e3861f6ecf8d3648300c',
       'name': 'Starbucks',
       'location': {'address': '637 Kingston Rd.',
        'crossStreet': 'at Main St.',
        'lat': 43.67879837444001,
        'lng': -79.2980449760153,

In [18]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
    
    if len(categories_list) == 0:
        print('None')
        return None
    else:
        return categories_list[0]['name']

In [19]:
venues = response['response']['groups'][0]['items']

nearby_venues = json_normalize(venues)

filtered_columns = ['venue.name', 'venue.categories','venue.location.lat','venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis = 1)

nearby_venues.columns = [i.split('.')[-1] for i in nearby_venues.columns]

nearby_venues

name    categories        lat        lng
0            Starbucks   Coffee Shop  43.678798 -79.298045
1  Grover Pub and Grub           Pub  43.679181 -79.297215
2        Upper Beaches  Neighborhood  43.680563 -79.292869

In [20]:
print ('{} venues were returned by Foursquare'.format(nearby_venues.shape[0]))

3 venues were returned by Foursquare


In [35]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude',
                  'Venue Name',           
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return nearby_venues, venues_list

In [43]:
toronto_venues, toronto_list = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The 

In [46]:
print(toronto_venues.shape)
toronto_venues.head(20)

(1705, 7)


Neighborhood  Neighborhood Latitude  \
0                    The Beaches              43.676357   
1                    The Beaches              43.676357   
2                    The Beaches              43.676357   
3   The Danforth West, Riverdale              43.679557   
4   The Danforth West, Riverdale              43.679557   
5   The Danforth West, Riverdale              43.679557   
6   The Danforth West, Riverdale              43.679557   
7   The Danforth West, Riverdale              43.679557   
8   The Danforth West, Riverdale              43.679557   
9   The Danforth West, Riverdale              43.679557   
10  The Danforth West, Riverdale              43.679557   
11  The Danforth West, Riverdale              43.679557   
12  The Danforth West, Riverdale              43.679557   
13  The Danforth West, Riverdale              43.679557   
14  The Danforth West, Riverdale              43.679557   
15  The Danforth West, Riverdale              43.679557   
16  The Danforth West, Riverdale              43.679557   
17  The Danforth West, Riverdale              43.679557   
18  The Danforth West, Riverdale              43.679557   
19  The Danforth West, Riverdale              43.679557   

    Neighborhood Longitude                          Venue Name  \
0               -79.293031                           Starbucks   
1               -79.293031                 Grover Pub and Grub   
2               -79.293031                       Upper Beaches   
3               -79.352188                            Pantheon   
4               -79.352188                        Dolce Gelato   
5               -79.352188                       MenEssentials   
6               -79.352188             Messini Authentic Gyros   
7               -79.352188                     Cafe Fiorentina   
8               -79.352188                               Mezes   
9               -79.352188         Christina's On The Danforth   
10              -79.352188                          La Diperie   
11              -79.352188  The Big Carrot Natural Food Market   
12              -79.352188                   The Auld Spot Pub   
13              -79.352188              Louis Cifer Brew Works   
14              -79.352188                           7 Numbers   
15              -79.352188                Moksha Yoga Danforth   
16              -79.352188                 Valley Farm Produce   
17              -79.352188                   Pizzeria Libretto   
18              -79.352188                              Ouzeri   
19              -79.352188           Astoria Shish Kebob House   

    Venue Latitude  Venue Longitude           Venue Category  
0        43.678798       -79.298045              Coffee Shop  
1        43.679181       -79.297215                      Pub  
2        43.680563       -79.292869             Neighborhood  
3        43.677621       -79.351434         Greek Restaurant  
4        43.677773       -79.351187           Ice Cream Shop  
5        43.677820       -79.351265           Cosmetics Shop  
6        43.677827       -79.350569         Greek Restaurant  
7        43.677743       -79.350115       Italian Restaurant  
8        43.677962       -79.350196         Greek Restaurant  
9        43.678240       -79.349185         Greek Restaurant  
10       43.677530       -79.352295           Ice Cream Shop  
11       43.677631       -79.353076        Health Food Store  
12       43.677335       -79.353130                      Pub  
13       43.677663       -79.351313                  Brewery  
14       43.677062       -79.353934       Italian Restaurant  
15       43.677622       -79.352116              Yoga Studio  
16       43.677999       -79.349969  Fruit & Vegetable Store  
17       43.678489       -79.347576              Pizza Place  
18       43.678193       -79.348908         Greek Restaurant  
19       43.677689       -79.351892         Greek Restaurant

Let's check how many venues returned by each group

In [48]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            53   
Brockton, Exhibition Place, Parkdale Village                           21   
Business reply mail Processing Centre969 Eastern                       17   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            48   
Central Bay Street                                                     82   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               16   
Church and Wellesley                                                   88   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             38   
Davisville North                                                        8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     15   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           19   
First Canadian Place, Underground city                                100   
Forest Hill North, Forest Hill West                                     4   
Harbord, University of Toronto                                         36   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                                              50   
High Park, The Junction South                                          24   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               65   
Moore Park, Summerhill East                                             4   
North Toronto West                                                     20   
Parkdale, Roncesvalles                                                 14   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     36   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        92   
Studio District                                                        41   
The Annex, North Midtown, Yorkville                                    25   
The Beaches                                                             3   
The Beaches West, India Bazaar                                         20   
The Danforth West, Riverdale                                           42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             53   
Brockton, Exhibition Place, Parkdale Village                            21   
Business reply mail Processing Centre969 Eastern                        17   
CN Tower, Bathurst Quay, Island airport, Harbou...                      14   
Cabbagetown, St. James Town                                             48   
Central Bay Street                                                      82   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                16   
Church and Wellesley                                                    88   
Commerce Court, Victoria Hotel     

How many unique categories can be curated from returned venues

In [52]:
print('There are {} unique venues'.format(len(toronto_venues['Venue Category'].unique())))

There are 234 unique venues


In [119]:
toronto_venues[toronto_venues['Venue Category'] == 'Neighborhood']

Neighborhood  Neighborhood Latitude  \
2                                          The Beaches              43.676357   
74                                     Studio District              43.659526   
733                           Adelaide, King, Richmond              43.650571   
820  Harbourfront East, Toronto Islands, Union Station              43.640816   

     Neighborhood Longitude        Venue Name  Venue Latitude  \
2                -79.293031     Upper Beaches       43.680563   
74               -79.340923       Leslieville       43.662070   
733              -79.384568  Downtown Toronto       43.653232   
820              -79.381752      Harbourfront       43.639526   

     Venue Longitude Venue Category  
2         -79.292869   Neighborhood  
74        -79.337856   Neighborhood  
733       -79.385296   Neighborhood  
820       -79.380688   Neighborhood

## 4.Analyze each neighborhood

In [106]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix = "", prefix_sep = "")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood']

# move neighborhood column to the first column
index = int(np.where(toronto_onehot.columns == 'Neighborhood')[0])
fixed_columns = [toronto_onehot.columns[index]] + list(toronto_onehot.columns[:index]) + list(toronto_onehot.columns[index + 1:])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  \
0                   The Beaches                  0               0   
1                   The Beaches                  0               0   
2                   The Beaches                  0               0   
3  The Danforth West, Riverdale                  0               0   
4  The Danforth West, Riverdale                  0               0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal     ...       \
0               0                0                 0     ...        
1               0                0                 0     ...        
2               0                0                 0     ...        
3               0                0                 0     ...        
4               0                0                 0     ...        

   Thrift / Vintage Store  Toy / Game Store  Trail  Train Station  \
0                       0                 0      0              0   
1                       0                 0      0              0   
2                       0                 0      0              0   
3                       0                 0      0              0   
4                       0                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Women's Store  Yoga Studio  
0         0              0            0  
1         0              0            0  
2         0              0            0  
3         0              0            0  
4         0              0            0  

[5 rows x 234 columns]

Examine dataframe size

In [97]:
toronto_onehot.shape

(1705, 234)

We have 234 columns which is strange because our previous number of unique venue categories was 234. We should have 235 columns (since we add 1 more column of the Neighborhood). This means that Neighborhood is actually a unique venue category. We can check this.

In [120]:
toronto_venues[toronto_venues['Venue Category'] == 'Neighborhood']

Neighborhood  Neighborhood Latitude  \
2                                          The Beaches              43.676357   
74                                     Studio District              43.659526   
733                           Adelaide, King, Richmond              43.650571   
820  Harbourfront East, Toronto Islands, Union Station              43.640816   

     Neighborhood Longitude        Venue Name  Venue Latitude  \
2                -79.293031     Upper Beaches       43.680563   
74               -79.340923       Leslieville       43.662070   
733              -79.384568  Downtown Toronto       43.653232   
820              -79.381752      Harbourfront       43.639526   

     Venue Longitude Venue Category  
2         -79.292869   Neighborhood  
74        -79.337856   Neighborhood  
733       -79.385296   Neighborhood  
820       -79.380688   Neighborhood

By looking up these venues on Foursquare web page, we found that they are actually a type of venue. So we will change the name of the Neighborhood name column to nbh_name.

In [122]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix = "", prefix_sep = "")

# add neighborhood column back to dataframe
toronto_onehot['nbh_name'] = toronto_venues['Neighborhood']

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

nbh_name  Accessories Store  Adult Boutique  \
0                   The Beaches                  0               0   
1                   The Beaches                  0               0   
2                   The Beaches                  0               0   
3  The Danforth West, Riverdale                  0               0   
4  The Danforth West, Riverdale                  0               0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal     ...       \
0               0                0                 0     ...        
1               0                0                 0     ...        
2               0                0                 0     ...        
3               0                0                 0     ...        
4               0                0                 0     ...        

   Thrift / Vintage Store  Toy / Game Store  Trail  Train Station  \
0                       0                 0      0              0   
1                       0                 0      0              0   
2                       0                 0      0              0   
3                       0                 0      0              0   
4                       0                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Women's Store  Yoga Studio  
0         0              0            0  
1         0              0            0  
2         0              0            0  
3         0              0            0  
4         0              0            0  

[5 rows x 235 columns]

Examine the dataframe

In [123]:
toronto_onehot.shape

(1705, 235)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [127]:
toronto_grouped = toronto_onehot.groupby('nbh_name').mean().reset_index()
toronto_grouped

nbh_name  Accessories Store  \
0                            Adelaide, King, Richmond               0.01   
1                                         Berczy Park               0.00   
2        Brockton, Exhibition Place, Parkdale Village               0.00   
3    Business reply mail Processing Centre969 Eastern               0.00   
4   CN Tower, Bathurst Quay, Island airport, Harbo...               0.00   
5                         Cabbagetown, St. James Town               0.00   
6                                  Central Bay Street               0.00   
7           Chinatown, Grange Park, Kensington Market               0.00   
8                                            Christie               0.00   
9                                Church and Wellesley               0.00   
10                     Commerce Court, Victoria Hotel               0.00   
11                                         Davisville               0.00   
12                                   Davisville North               0.00   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...               0.00   
14           Design Exchange, Toronto Dominion Centre               0.00   
15                       Dovercourt Village, Dufferin               0.00   
16             First Canadian Place, Underground city               0.00   
17                Forest Hill North, Forest Hill West               0.00   
18                     Harbord, University of Toronto               0.00   
19  Harbourfront East, Toronto Islands, Union Station               0.00   
20                          Harbourfront, Regent Park               0.00   
21                      High Park, The Junction South               0.00   
22                                      Lawrence Park               0.00   
23                           Little Portugal, Trinity               0.00   
24                        Moore Park, Summerhill East               0.00   
25                                 North Toronto West               0.00   
26                             Parkdale, Roncesvalles               0.00   
27                                           Rosedale               0.00   
28                                           Roselawn               0.00   
29                                 Runnymede, Swansea               0.00   
30                           Ryerson, Garden District               0.00   
31                                     St. James Town               0.01   
32                    Stn A PO Boxes 25 The Esplanade               0.00   
33                                    Studio District               0.00   
34                The Annex, North Midtown, Yorkville               0.00   
35                                        The Beaches               0.00   
36                     The Beaches West, India Bazaar               0.00   
37                       The Danforth West, Riverdale               0.00   

    Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0         0.000000           0.000000  0.000000            0.000000   
1         0.000000           0.000000  0.000000            0.000000   
2         0.000000           0.000000  0.000000            0.000000   
3         0.000000           0.000000  0.000000            0.000000   
4         0.000000           0.000000  0.071429            0.071429   
5         0.000000           0.000000  0.000000            0.000000   
6         0.000000           0.000000  0.000000            0.000000   
7         0.000000           0.000000  0.000000            0.000000   
8         0.000000           0.000000  0.000000            0.000000   
9         0.011364           0.011364  0.000000            0.000000   
10        0.000000           0.000000  0.000000            0.000000   
11        0.000000           0.000000  0.000000            0.000000   
12        0.000000           0.000000  0.000000            0.000000   
13        0.000000           0.000000  0.000000            0.000000   
14        0.000

#### Confirm the new size

In [132]:
print(toronto_grouped.shape)
toronto_grouped.shape[0] == len(toronto_data)

(38, 235)


True

In [156]:
temp = toronto_grouped[toronto_grouped['nbh_name'] == 'Berczy Park'].T.reset_index()
temp = temp.iloc[1:]
temp.columns = ['venue','freq']
temp['freq'] = temp['freq'].astype(float)
temp = temp.round({'freq' : 2})
print(temp.sort_values(by  = 'freq', ascending = False).reset_index(drop = True))

                           venue  freq
0                    Coffee Shop  0.09
1                   Cocktail Bar  0.06
2                     Steakhouse  0.04
3                     Restaurant  0.04
4                       Beer Bar  0.04
5                           Café  0.04
6                    Cheese Shop  0.04
7             Seafood Restaurant  0.04
8                         Bakery  0.04
9                 Farmers Market  0.04
10                  Burger Joint  0.02
11                 Shopping Mall  0.02
12                Clothing Store  0.02
13       Comfort Food Restaurant  0.02
14                  Concert Hall  0.02
15                Cosmetics Shop  0.02
16                        Bistro  0.02
17                           Pub  0.02
18                         Diner  0.02
19                          Park  0.02
20                   Fish Market  0.02
21                      Fountain  0.02
22             French Restaurant  0.02
23                        Museum  0.02
24              Greek Res

In [163]:
num_top_venues = 5

for hood in toronto_grouped['nbh_name']:
    print("-----"+hood+"-----")
    temp = toronto_grouped[toronto_grouped['nbh_name'] == hood].T.reset_index()
    temp = temp.iloc[1:]
    temp.columns = ['venue','freq']
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq':2})
    print(temp.sort_values('freq', ascending = False).reset_index(drop = True).head(num_top_venues))
    print()

-----Adelaide, King, Richmond-----
                 venue  freq
0          Coffee Shop  0.07
1                 Café  0.06
2  American Restaurant  0.04
3      Thai Restaurant  0.04
4           Steakhouse  0.04

-----Berczy Park-----
          venue  freq
0   Coffee Shop  0.09
1  Cocktail Bar  0.06
2    Steakhouse  0.04
3    Restaurant  0.04
4      Beer Bar  0.04

-----Brockton, Exhibition Place, Parkdale Village-----
                  venue  freq
0           Coffee Shop  0.14
1        Breakfast Spot  0.10
2                  Café  0.10
3         Burrito Place  0.05
4  Caribbean Restaurant  0.05

-----Business reply mail Processing Centre969 Eastern-----
              venue  freq
0       Yoga Studio  0.06
1     Auto Workshop  0.06
2        Comic Shop  0.06
3              Park  0.06
4  Recording Studio  0.06

-----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara-----
              venue  freq
0   Airport Service  0.14
1  Airport Ter

#### Put that into a pandas dataframe

Write a function to sort in descending order

In [164]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_sorted = row_categories.sort_values(ascending = False)
    return row_sorted.index.values[0:num_top_venues]

Create a dataframe with ten most common venues for each neighborhood

In [177]:
num_top_venues = 10

columns = ['nbh_name']
indicator = ['st','nd','rd']

for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicator[id]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

nbh_venues_sorted = pd.DataFrame(columns = columns)
nbh_venues_sorted['nbh_name'] = toronto_grouped['nbh_name']
for i in range(0, len(toronto_grouped)):
    nbh_venues_sorted.iloc[i, 1:] = return_most_common_venues(toronto_grouped.iloc[i,:], num_top_venues)
                       

In [178]:
nbh_venues_sorted

nbh_name 1th Most Common Venue  \
0                            Adelaide, King, Richmond           Coffee Shop   
1                                         Berczy Park           Coffee Shop   
2        Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
3    Business reply mail Processing Centre969 Eastern           Yoga Studio   
4   CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   
5                         Cabbagetown, St. James Town           Coffee Shop   
6                                  Central Bay Street           Coffee Shop   
7           Chinatown, Grange Park, Kensington Market                  Café   
8                                            Christie         Grocery Store   
9                                Church and Wellesley   Japanese Restaurant   
10                     Commerce Court, Victoria Hotel           Coffee Shop   
11                                         Davisville        Sandwich Place   
12                                   Davisville North          Burger Joint   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...           Coffee Shop   
14           Design Exchange, Toronto Dominion Centre           Coffee Shop   
15                       Dovercourt Village, Dufferin                Bakery   
16             First Canadian Place, Underground city           Coffee Shop   
17                Forest Hill North, Forest Hill West                  Park   
18                     Harbord, University of Toronto                  Café   
19  Harbourfront East, Toronto Islands, Union Station           Coffee Shop   
20                          Harbourfront, Regent Park           Coffee Shop   
21                      High Park, The Junction South                   Bar   
22                                      Lawrence Park                  Park   
23                           Little Portugal, Trinity                   Bar   
24                        Moore Park, Summerhill East            Restaurant   
25                                 North Toronto West        Clothing Store   
26                             Parkdale, Roncesvalles             Gift Shop   
27                                           Rosedale                  Park   
28                                           Roselawn        Ice Cream Shop   
29                                 Runnymede, Swansea           Coffee Shop   
30                           Ryerson, Garden District           Coffee Shop   
31                                     St. James Town           Coffee Shop   
32                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
33                                    Studio District                  Café   
34                The Annex, North Midtown, Yorkville           Coffee Shop   
35                                        The Beaches           Coffee Shop   
36                     The Beaches West, India Bazaar                  Park   
37                       The Danforth West, Riverdale      Greek Restaurant   

            2th Most Common Venue 3th Most Common Venue 4th Most Common Venue  \
0                            Café   American Restaurant            Steakhouse   
1                    Cocktail Bar    Seafood Restaurant                Bakery   
2                            Café        Breakfast Spot         Grocery Store   
3                   Auto Workshop            Comic Shop           Pizza Place   
4                 Airport Service      Airport Terminal       Harbor / Marina   
5                      Restaurant           Pizza Place    Italian Restaurant   
6                            Café    Italian Restaurant        Sandwich Place   
7   Vegetarian / Vegan Restaurant    Chinese Restaurant                   Bar   
8                            Café                  Park             Nightclub   
9                     Coffee Shop               Gay Bar          Burger Joint   
10                          Hotel                  Café            Resta

## 5.Cluster neighborhoods

Run k-mean to cluster neighborhoods into 5 clusters

In [257]:
#set number of clusters 
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('nbh_name', axis = 1)

#run k-mean clustering
kmeans = KMeans(n_clusters = kclusters, random_state = 0).fit(toronto_grouped_clustering)

#check cluster label
kmeans.labels_

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 1, 2, 2, 2, 2,
       4, 2, 1, 2, 2, 1, 3, 2, 2, 2, 2, 2, 2, 0, 2, 2])

Create a new dataframe that include the cluster as well as most common venues

In [258]:
toronto_merged = toronto_data

toronto_merged['Cluster Labels'] = kmeans.labels_

toronto_merged = toronto_merged.join(nbh_venues_sorted.set_index('nbh_name'), on = toronto_merged['Neighborhood'])

In [259]:
toronto_merged

PostalCode           Borough  \
0         M4E      East Toronto   
1         M4K      East Toronto   
2         M4L      East Toronto   
3         M4M      East Toronto   
4         M4N   Central Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
10        M4W  Downtown Toronto   
11        M4X  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
14        M5B  Downtown Toronto   
15        M5C  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
18        M5H  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
21        M5L  Downtown Toronto   
22        M5N   Central Toronto   
23        M5P   Central Toronto   
24        M5R   Central Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
27        M5V  Downtown Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
30        M6G  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
34        M6P      West Toronto   
35        M6R      West Toronto   
36        M6S      West Toronto   
37        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                                         The Beaches  43.676357 -79.293031   
1                        The Danforth West, Riverdale  43.679557 -79.352188   
2                      The Beaches West, India Bazaar  43.668999 -79.315572   
3                                     Studio District  43.659526 -79.340923   
4                                       Lawrence Park  43.728020 -79.388790   
5                                    Davisville North  43.712751 -79.390197   
6                                  North Toronto West  43.715383 -79.405678   
7                                          Davisville  43.704324 -79.388790   
8                         Moore Park, Summerhill East  43.689574 -79.383160   
9   Deer Park, Forest Hill SE, Rathnelly, South Hi...  43.686412 -79.400049   
10                                           Rosedale  43.679563 -79.377529   
11                        Cabbagetown, St. James Town  43.667967 -79.367675   
12                               Church and Wellesley  43.665860 -79.383160   
13                          Harbourfront, Regent Park  43.654260 -79.360636   
14                           Ryerson, Garden District  43.657162 -79.378937   
15                                     St. James Town  43.651494 -79.375418   
16                                        Berczy Park  43.644771 -79.373306   
17                                 Central Bay Street  43.657952 -79.387383   
18                           Adelaide, King, Richmond  43.650571 -79.384568   
19  Harbourfront East, Toronto Islands, Union Station  43.640816 -79.381752   
20           Design Exchange, Toronto Dominion Centre  43.647177 -79.381576   
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
22                                           Roselawn  43.711695 -79.416936   
23                Forest Hill North, Forest Hill West  43.696948 -79.411307   
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678   
25                     Harbord, University of Toronto  43.662696 -79.400049   
26          Chinatown, Grange Park, Kensington Market  43.653206 -79.400049   
27  CN Tower, Bathurst Quay, Island airport, Harbo...  43.628947 -79.394420   
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
29             First Canadian Place, Underground city  43.648429 -79.382280   
30                                           Christie  43.669542 -79.422564   
31                       Dovercourt Village, Dufferin  43.669005 -79.442259   
32                           L

Finally, let's visualize the resulting clusters

In [260]:
#create map
map_clusters = folium.Map(location = [latitude, longitude], zoom_start = 10)

#set color scheme for clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 6. Examine Clusters

Examine each cluster and determine the discriminating venue categories that distinguish each cluster.

#### Cluster 1

In [263]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1th Most Common Venue 2th Most Common Venue  \
13  Harbourfront, Regent Park           Coffee Shop                Bakery   
35     Parkdale, Roncesvalles             Gift Shop        Breakfast Spot   

   3th Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
13                  Park                  Café                   Restaurant   
35             Piano Bar      Cuban Restaurant  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
13                   Pub    Mexican Restaurant        Breakfast Spot   
35                   Bar               Dog Run          Burger Joint   

   9th Most Common Venue 10th Most Common Venue  
13               Theater      Health Food Store  
35          Dessert Shop             Restaurant

#### Cluster 2

In [264]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1th Most Common Venue  \
17                                 Central Bay Street           Coffee Shop   
24                The Annex, North Midtown, Yorkville           Coffee Shop   
27  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

   2th Most Common Venue 3th Most Common Venue 4th Most Common Venue  \
17                  Café    Italian Restaurant        Sandwich Place   
24                  Café        Sandwich Place           Pizza Place   
27       Airport Service      Airport Terminal       Harbor / Marina   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17                   Bar       Bubble Tea Shop          Burger Joint   
24     French Restaurant        Cosmetics Shop                   Pub   
27         Boat or Ferry               Airport    Airport Food Court   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17   Japanese Restaurant        Ice Cream Shop     Falafel Restaurant  
24          Burger Joint         Metro Station           Liquor Store  
27          Airport Gate      Sculpture Garden               Boutique

#### Cluster 3

In [265]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1th Most Common Venue  \
0                                         The Beaches           Coffee Shop   
1                        The Danforth West, Riverdale      Greek Restaurant   
2                      The Beaches West, India Bazaar                  Park   
3                                     Studio District                  Café   
4                                       Lawrence Park                  Park   
5                                    Davisville North          Burger Joint   
6                                  North Toronto West        Clothing Store   
7                                          Davisville        Sandwich Place   
8                         Moore Park, Summerhill East            Restaurant   
9   Deer Park, Forest Hill SE, Rathnelly, South Hi...           Coffee Shop   
10                                           Rosedale                  Park   
11                        Cabbagetown, St. James Town           Coffee Shop   
12                               Church and Wellesley   Japanese Restaurant   
14                           Ryerson, Garden District           Coffee Shop   
15                                     St. James Town           Coffee Shop   
16                                        Berczy Park           Coffee Shop   
18                           Adelaide, King, Richmond           Coffee Shop   
19  Harbourfront East, Toronto Islands, Union Station           Coffee Shop   
20           Design Exchange, Toronto Dominion Centre           Coffee Shop   
21                     Commerce Court, Victoria Hotel           Coffee Shop   
23                Forest Hill North, Forest Hill West                  Park   
25                     Harbord, University of Toronto                  Café   
26          Chinatown, Grange Park, Kensington Market                  Café   
29             First Canadian Place, Underground city           Coffee Shop   
30                                           Christie         Grocery Store   
31                       Dovercourt Village, Dufferin                Bakery   
32                           Little Portugal, Trinity                   Bar   
33       Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
34                      High Park, The Junction South                   Bar   
36                                 Runnymede, Swansea           Coffee Shop   
37   Business reply mail Processing Centre969 Eastern           Yoga Studio   

            2th Most Common Venue 3th Most Common Venue 4th Most Common Venue  \
0                             Pub          Neighborhood           Yoga Studio   
1                     Coffee Shop        Ice Cream Shop             Bookstore   
2                       Pet Store            Steakhouse         Burrito Place   
3                     Coffee Shop   American Restaurant                Bakery   
4                     Swim School              Bus Line    Dim Sum Restaurant   
5                         Dog Run                  Park           Pizza Place   
6             Sporting Goods Shop           Coffee Shop           Yoga Studio   
7                     Pizza Place          Dessert Shop    Seafood Restaurant   
8                            Park            Playground          Tennis Court   
9               Convenience Store                   Pub      Sushi Restaurant   
10                     Playground                 Trail           Yoga Studio   
11                     Restaurant           Pizza Place    Italian Restaurant   
12                    Coffee Shop               Gay Bar          Burger Joint   
14                 Clothing Store                  Café        Cosmetics Shop   
15                           Café            Restaurant        Clothing Store   
16                   Cocktail Bar    Seafood Restaurant                Bakery   
18                           Café   American Restaurant            Steakhouse   
19                          Hotel              Aquariu

#### Cluster 4

In [266]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1th Most Common Venue  \
28  Stn A PO Boxes 25 The Esplanade           Coffee Shop   

   2th Most Common Venue 3th Most Common Venue 4th Most Common Venue  \
28                  Café              Beer Bar                 Hotel   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
28    Seafood Restaurant            Restaurant          Cocktail Bar   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
28          Burger Joint   Japanese Restaurant         Farmers Market

#### Cluster 5

In [267]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1th Most Common Venue 2th Most Common Venue  \
22     Roselawn        Ice Cream Shop                Garden   

   3th Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
22           Yoga Studio          Dessert Shop    Falafel Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
22           Event Space  Ethiopian Restaurant     Electronics Store   

          9th Most Common Venue 10th Most Common Venue  
22  Eastern European Restaurant    Dumpling Restaurant